### ISSUE TITLE
Foxx feature request

### ISSUE THREAD

**Issue by ansarizafar** [arangodb/arangodb#2419]
Foxx is a great idea but need improvement and more love.  Please consider to provide following features 
1. Full NodeJs development environment without current limitations for Foxx services.
2. Stored procedures, which are only accessible from outside with database credentials and not on a publicly available route.   
3. Only Endpoint code should execute and not the whole service code on each request.
4. Code hot reload when service is in development mode. 
5. New Logo.

**Comment by dothebart** [comment]
1) duplicate of #1538 ; While some more extensions may become available in the future full compatibility will never be possible - nodejs is a to fast moving target for that.
2) By design foxx apps have to handle their authentication on their own, or as microservices do, utilize another service.
3) matching of rest namespaces etc. has to be done on each request. 
4) is what the development mode is actually about.
5) we especially like our current logo.

Closing.

**Comment by weinberger** [comment]
@ansarizafar 
I would love to understand you points a little bit more in detail. 
ad 1) ArangoDB and NodeJs have different approaches. Foxx isn't a competitor.
ad 2) Did you have see the user defined functions?
ad 3) I'm not sure if I got your point. Can you go a little bit more in detail?
ad 5) That's really a personal thing. Why do you think we need a new logo?

**Comment by ansarizafar** [comment]
@weinberger 
Yes I know Foxx is similar and different than Nodejs in many ways. The stated goal of Foxx framework is to allow application developers to write their data access and domain logic either as REST endpoints or entire standalone applications running directly within the database with native access to in-memory data. The main advantages of this approach are 
Less network overhead/round trips. 
Easy distribution and version control.

Foxx in current form also has some disadvantages.
Each Foxx service/app is mounted on an endpoint and inside each service/app we can define our own endpoints by using Foxx router module. Unlike Nodejs apps, When a client sends a request to any endpoint, (defined with in the service/app) the whole service/app code re-executes (not just endpoint code) on each request as in Foxx matching of rest namespaces etc. has to be done on each request. If  I am writing a micro service mounted on a single endpoint with Foxx then its ok to re-execute whole service code but If I am writing and app with multiple endpoints (defined within app) then this behavior is not efficient. Ideally, like Nodejs, app initialization code must run only once (when app is deployed) and only the code attached to the requested endpoint should re-execute on each request.

Foxx should work as much as possible like Nodejs, specially when a new rocksdb based engine will be available with next Arangodb release. 

The ability to disable Foxx framework (If not using it) would also be useful to conserve system resources.

Yes, I have seen user defined functions but since they won't be able to utilize indices they are not an alternative of stored procedure. In my opinion Arangodb must also provide the ability to write and store stored procedures within database that are only accessible with database credentials via arangodb clients and not on a publicly available route.

Java driver supports Velocypack. Please also provide  velocypack support with araangodb nodejs client, Foxx framework and admin ui should provide support for date and number types provided by velocypack

Personally I don't like hollow logos.

**Comment by pluma** [comment]
> Unlike Nodejs apps, When a client sends a request to any endpoint, (defined with in the service/app) the whole service/app code re-executes (not just endpoint code) on each request as in Foxx matching of rest namespaces etc. has to be done on each request.

Services are indeed initialised lazily (i.e. the first time a request is made to a service). However unless you're in development mode or have installed/replaced/updated the services in between requests, the service should not need to be re-initialised for each consecutive request.

One reason you might run into behaviour that _looks_ like the service is initialised for every request is that ArangoDB is multi-threaded and Foxx runs in multiple isolated contexts (one per thread). Because V8 is single-threaded and these contexts are completely isolated, the service will need to be initialised in each context separately and because this happens lazily, you might see the service be initialised every time if you make as many requests as you have contexts and happen to hit a fresh context every time. However this again would only occur on the first request handled by a service in the given context.

> Code hot reload

This is basically the experience you get with development mode: when a service runs in development mode, the service is re-initialised from disk every time it handles a new request. If you don't want to work on the server filesystem directly, you can use conventional tools like lsyncd to continuously sync your local working copy to the deployed service and have the changes picked up during development mode.

> Personally I don't like hollow logos.

I suspect you're thinking of the ASCII fox logo. We actually do have a new Foxx logo (a cartoon fox), which is also used on the ArangoDB website and in marketing materials.

**Comment by MrPieces** [comment]
ansarizafar took the liberty to close this issue. But please feel free to reopen it if you have further feedback. It's always super helpful for us to get your input.
